Update support page and enable custom links. #6493

Shavindra opened the pull request:
<!-- This template is for Devs to give QA details before moving the issue To-Test -->
### Summary
Fixes 
 - #6493 -Update Home pages
 - #6294 - Add ability to customise the docs link.
  
 <!-- Define findings related to the feature or bug issue. -->

### Areas or cases that should be tested

1. Navigate to Configuration > Global Settings > Links
2. Add Custom links
3. Apply 
4. Navigate back to homepage and it should have newly added custom links.

### Screenshot/Video
![Uploading image.png…]()

richard-cox commented on shell/pages/c/_cluster/settings/links.vue:
```diff
@@ -0,0 +1,203 @@
+      </label>
+      <!-- Incremental Loading -->
+      <div class="ui-links-setting">
+        <KeyValue
```
Should the user be able to add more entries to this collection via the  `Add Label` button?

richard-cox commented on shell/pages/c/_cluster/settings/links.vue:
```diff
@@ -0,0 +1,203 @@
+  },
+  {
+    order: 5,
+    key:   'Getting Started',
```
as mentioned above, these shouldn't be en-us but a key that can be used to lookup the values in l10n files

richard-cox commented on shell/components/CommunityLinks.vue:
```diff
@@ -66,11 +82,11 @@ export default {
+        <a :href="value" target="_blank" rel="noopener noreferrer nofollow"> {{ name }} </a>
       </div>
-
+      <slot />
```
![image](https://user-images.githubusercontent.com/18697775/189913654-814aa2d6-f858-4db6-a8f1-4c50e7ad0694.png)

There's a missmatch between the links in the support page (are external and have an icon to show it) and the external one's from options (which don't have the icon)

Shavindra replied:
Discussed this with Ken. All links now would open from a new window. The icon will appear next to the title. `SimpleBox` updated to take a slot now.

richard-cox commented on shell/pages/c/_cluster/settings/links.vue:
```diff
@@ -0,0 +1,203 @@
+    <div v-if="mode === 'edit'" class="mt-20">
+      <AsyncButton class="pull-right" mode="apply" @click="save" />
+      <button class="pull-right btn role-secondary mr-10" @click="useDefaults">
+        Restore default
```
l10n

richard-cox commented on shell/pages/c/_cluster/settings/links.vue:
```diff
@@ -0,0 +1,203 @@
+        <KeyValue
+          v-model="value"
+          :as-map="false"
+          :key-label="'text'"
```
l10n, should probably be `Link Label`

richard-cox commented on shell/pages/c/_cluster/settings/links.vue:
```diff
@@ -0,0 +1,203 @@
+          v-model="value"
+          :as-map="false"
+          :key-label="'text'"
+          :value-label="'URL'"
```
l10n

richard-cox commented on shell/pages/c/_cluster/settings/links.vue:
```diff
@@ -0,0 +1,203 @@
+      <label class="text-label">
+        {{ t(`customLinks.description`, {}, true) }}
+      </label>
+      <!-- Incremental Loading -->
```
odd comment?

richard-cox commented on shell/pages/c/_cluster/settings/links.vue:
```diff
@@ -0,0 +1,203 @@
+      </label>
+      <!-- Incremental Loading -->
+      <div class="ui-links-setting">
+        <KeyValue
```
should the `text` column be editable?

Shavindra replied:
User can edit the key and value - fully customisable

richard-cox commented on shell/pages/c/_cluster/settings/links.vue:
```diff
@@ -0,0 +1,203 @@
+      this.value = [...DEFAULT_CUSTOM_LINKS, DEFAULT_SUPPORT_LINK, ...nonCommercialRancherLinks];
+    },
+
+    cancel() {
```
is this used anywhere?

richard-cox commented on shell/pages/c/_cluster/settings/links.vue:
```diff
@@ -0,0 +1,203 @@
+        this.uiCommunitySetting.value = '';
+
+        return Promise.all([
+          this.uiIssuesSetting.save(),
```
this is the same function called twice

richard-cox commented on shell/pages/c/_cluster/settings/links.vue:
```diff
@@ -0,0 +1,203 @@
+
+    deprecateIssueLinks() {
+      if (this.uiIssuesSetting.value || this.uiIssuesSetting.value) {
+        this.uiIssuesSetting.value = '';
```
These values should be migrated to their new setting before they are cleared

Shavindra replied:
It happens here

https://github.com/rancher/dashboard/pull/6720/files#diff-e7649644ab782f40aabab49a017a95eeedaf9f9456abce5b5d2315db353f3a3cR73-R75

richard-cox reviewed (changes requested):
- The items in the global settings menu should probably be ordered somehow or, where appropriate, put into a group. At the moment they seem a bit messy. Check with Ken. `Branding`, `banners` and `home links` are all visual UI things. `Performance` is functional UI. `Feature Flags` is functional Rancher manager (including UI). `Settings` are all over the place. 
  - ![image](https://user-images.githubusercontent.com/18697775/189908170-677a2820-a643-4f39-bbad-318467967818.png)
- ~should the ui issues and community link options be removed from the Branding page or marked as deprecated?~ Think one of the issues says these should be removed
  - ![image](https://user-images.githubusercontent.com/18697775/189914783-f7114a47-735f-421f-8b45-3b57c54df07e.png)
- We should keep a setting that will remove the community supported based links. Check with Neil
- Links should probably be alphabetically ordered
  - ![image](https://user-images.githubusercontent.com/18697775/189918667-1efbc55a-33a3-40ee-b6d6-1ca6d3f37b75.png)

In addition

- #6493
  - ![image](https://user-images.githubusercontent.com/18697775/189918326-8c6f9900-6028-4360-9b10-46995c34d478.png) Think this should be removed as well? Check with Neil
- #6663
  - The 'add subscription' button is still visible

richard-cox commented on shell/components/CommunityLinks.vue:
```diff
@@ -21,12 +23,27 @@ export default {
+    }
+
+    // Fallback:
+    // NB: this.uiIssueSetting is deprecated now.
```
Update: I understand what's happening now. We probably should add an update to the backend settings over in `rancher/rancher` to state it's deprecated

Shavindra replied:
Yes, its deprecated

richard-cox replied:
For reference this is the comment that needs updating https://github.com/rancher/rancher/blob/release/v2.7/pkg/settings/setting.go#L176

Shavindra replied:
https://github.com/rancher/rancher/pull/38910

richard-cox commented on shell/config/footer.js:
```diff
@@ -1,19 +1,27 @@
+    Docs:              DOCS_BASE,
+    Forums:            'https://forums.rancher.com/',
+    Slack:             'https://slack.rancher.io',
+    'File an Issue':   issues,
```
Update: Was hunting for this one earlier, ignore. There's another comment that requests these are keys that relate to l10n entries

Shavindra replied:
I wasn't quite happy with this per se. User can edit the key and value. I can move it to yaml. However, I don't think it would get used?

richard-cox replied:
Setting the links to be keyed by l10n key as before would work in a similar way. 

The `a` display label would use the link's key to look up the translation. As long as they were sensibly named the KeyValue component could continue to use the key in the name column

richard-cox replied:
`'File an issue'` and `'Commercial Support'` need updating.

Importantly the keys from `DEFAULT_CUSTOM_LINKS` (and `DEFAULT_SUPPORT_LINK` `COMMUNITY_LINKS`) need updating

richard-cox commented on shell/pages/support/index.vue:
```diff
@@ -104,11 +104,12 @@ export default {
 
     options() {
-      return options( this.uiIssuesSetting?.value, this.communitySetting?.value === 'false');
+      return options(true, this.uiIssuesSetting?.value, this.communitySetting?.value === 'false');
```
Update: I was getting confused between hasSupport and isSupport

Shavindra commented on shell/components/form/KeyValue.vue:
```diff
@@ -222,69 +222,69 @@ export default {
-  data() {
-    const rows = [];
+  computed: {
+    rows() {
```
Restoring to defaults button push the items in to array. It wasn't reactive.

Shavindra commented on shell/components/CommunityLinks.vue:
```diff
@@ -36,22 +53,21 @@ export default {
-        return options(this.uiIssuesSetting?.value, true);
+      // Custom links set from settings
+      if (this.uiCustomLinks) {
+        return JSON.parse(this.uiCustomLinks.value).reduce((prev, curr) => {
```
Resolved

Shavindra commented on shell/pages/home.vue:
```diff
@@ -118,6 +116,11 @@ export default {
 
+    isSupport() {
+      // NB: This is temporary until API implemented
+      return false;
```
No it should be part of my other. removed.

richard-cox reviewed (changes requested):
There are a number of unresolved comments, plus from the last review

- Comment 1 from https://github.com/rancher/dashboard/pull/6720#pullrequestreview-1105713714
- Comment 3 from https://github.com/rancher/dashboard/pull/6720#pullrequestreview-1105713714
- Comment 4 from https://github.com/rancher/dashboard/pull/6720#pullrequestreview-1105713714

- Addition Section - Comment 1 from https://github.com/rancher/dashboard/pull/6720#pullrequestreview-1105713714

Shavindra commented:
> * We should keep a setting that will remove the community supported based links. Check with Neil



> There are a number of unresolved comments, plus from the last review
> 
> * Comment 1 from [#6720 (review)](https://github.com/rancher/dashboard/pull/6720#pullrequestreview-1105713714) - Discussed with Ken. for now would keep as it is
> * Comment 3 from [#6720 (review)] (https://github.com/rancher/dashboard/pull/6720#pullrequestreview-1105713714) - Not quite sure what you mean. Once API is implemented based on that support link would be shown. There is only 2 either commercial link or non commercial link. This is awaiting API implementation for support.
> * Comment 4 from [#6720 (review)] (https://github.com/rancher/dashboard/pull/6720#pullrequestreview-1105713714) - Discussed with Ken. for now would keep as it is
> * Addition Section - Comment 1 from [#6720 (review)](https://github.com/rancher/dashboard/pull/6720#pullrequestreview-1105713714) - Discussed with Neil keeping this as it is.

richard-cox commented:
> > Comment 3 from [#6720 (review)](https://github.com/rancher/dashboard/pull/6720#pullrequestreview-1105713714)
> 
> Not quite sure what you mean. Once API is implemented based on that support link would be shown. There is only 2 either commercial link or non commercial link. This is awaiting API implementation for support.

Did the setting apply to the visibility of the whole card or just the two links? If the former, does removing all links also remove the card, if the later then we can drop that comment.

> > Addition Section - Comment 1 from https://github.com/rancher/dashboard/pull/6720#pullrequestreview-1105713714
> 
> Discussed with Neil keeping this as it is.

Might be worth commenting on the issue, to ensure QA don't also hit the same question

richard-cox commented on shell/utils/string.js:
```diff
@@ -300,3 +300,11 @@ export function isIpv4(ip) {
   return reg.test(ip);
 }
+
+export function translateKeysWithFallback(items, tFallbackFn, key = 'key') {
```
This would fit better in the i18n store, `multiWithFallback`

richard-cox commented on shell/utils/string.js:
```diff
@@ -300,3 +300,11 @@ export function isIpv4(ip) {
 }
+
+export function translateKeysWithFallback(items, tFallbackFn, key = 'key') {
+  return [...items].map((item) => {
```
No need to spread items before map

richard-cox commented on shell/pages/support/index.vue:
```diff
@@ -65,7 +64,6 @@ export default {
       supportSetting:       null,
       brandSetting:         null,
       uiIssuesSetting:      null,
-      communitySetting:     null,
```
`return options(true, this.uiIssuesSetting?.value, this.communitySetting?.value === 'false');` also needs updating

richard-cox reviewed (changes requested):
- New (undiscovered bug) - `Links` card should not show if there are no links. Applies to home page card and possibly support page card?
- l10n can be removed for `landing.commercial` now that it's part of the links section

There are also 4 unresolved comments (excluding the non-blocking rancher/rancher issue) and some failing gates.

Shavindra commented:
> * New (undiscovered bug) - `Links` card should not show if there are no links. Applies to home page card and possibly support page card?

This is hidden on the homepage only. Support page has 2 additional links that would always be shown

> * l10n can be removed for `landing.commercial` now that it's part of the links section

Removed.

richard-cox reviewed (changes requested):
There's a couple of issues with the 'no links' scenario
- Home page
  - The changes remove the card but keep the space the card was in. It looks like there was a mechanism to address this before (see home page `showSidePanel`). The `CommunityLinks` component probably needs to emit whether it has options or not... which can be plumbed in to `showSidePanel`
  - ![image](https://user-images.githubusercontent.com/18697775/191760198-4edc5259-c65e-4e35-b4cd-2af830cca29e.png)
- Support Page
  - Remove all links also removes the card, from your comment there should always be at least two there
  - ![image](https://user-images.githubusercontent.com/18697775/191760381-41a4c495-9754-4eef-9bb2-71c0e75c7c38.png)
- New comment
  - The 'you can change what you see' banner isn't aligned with the links card
  - ![image](https://user-images.githubusercontent.com/18697775/191761253-79c01cac-7575-4a77-be3f-4288e7ac91f4.png)
  - ![image](https://user-images.githubusercontent.com/18697775/191761323-b8821691-9642-417b-9efa-ed9dc1c9099e.png)
- There are also 4 unresolved comments (excluding the non-blocking rancher/rancher issue).

Shavindra commented:
> * The changes remove the card but keep the space the card was in. It looks like there was a mechanism to address this before (see home page `showSidePanel`). The `CommunityLinks` component probably needs to emit whether it has options or not... which can be plumbed in to `showSidePanel`

This is fixed with flexbox.

Shavindra commented:
> * The changes remove the card but keep the space the card was in. It looks like there was a mechanism to address this before (see home page `showSidePanel`). The `CommunityLinks` component probably needs to emit whether it has options or not... which can be plumbed in to `showSidePanel`

This is fixed with flexbox. 

> * Remove all links also removes the card, from your comment there should always be at least two there

Fixed

> * The 'you can change what you see' banner isn't aligned with the links card

Fixed

> * There are also 4 unresolved comments (excluding the non-blocking rancher/rancher issue).

Resolved

Shavindra commented on shell/components/CommunityLinks.vue:
```diff
@@ -66,11 +82,11 @@ export default {
+    <SimpleBox :title="t('customLinks.displayTitle')" :pref="pref" :pref-key="prefKey">
       <div v-for="(value, name) in options" :key="name" class="support-link">
-        <a v-t="name" :href="value" target="_blank" rel="noopener noreferrer nofollow" />
+        <a :href="value" target="_blank" rel="noopener noreferrer nofollow"> {{ name }} </a>
```
Discussed this with Ken. All links will open in a new tab.

richard-cox commented on shell/pages/home.vue:
```diff
@@ -287,9 +287,9 @@ export default {
       </div>
 
-      <div class="row">
-        <div :class="{'span-9': showSidePanel, 'span-12': !showSidePanel }" class="col">
```
If we're not using `showSidePanel` anymore the computed property can be removed. This also applies to `showCommercialSupport` and `showCommunityLinks`

richard-cox reviewed (changes requested):
- One minor comment regarding unused code
- I'm not quite sold on internal links opening in new tabs
  - https://github.com/rancher/dashboard/pull/6720#discussion_r969626881 and https://github.com/rancher/dashboard/pull/6720#discussion_r969630124
  - Neil has mentioned he'd generally like to take a look, so will bow to his opinion

Apart from those two above, LGTM

Shavindra commented:
> I'm not quite sold on internal links opening in new tabs

It would probably be an issue for `docs` link only. Others are not really internal links of the app. Also they are fully customisable. As it stand usage of external link icon is not consistent in the app.

@kwwii recommendation 
cc: @nwmac

richard-cox commented:
It would include both `Get Started`, `Commercial Support` and whatever links the user might add.

TBH we don't have to stick to the icon notifying an external link (i don't think they're used anywhere else), but adding quick links to internal pages that take the user automatically to a new tab is a bit clunky.

Shavindra commented:
> TBH we don't have to stick to the icon notifying an external link (i don't think they're used anywhere else), but adding quick links to internal pages that take the user automatically to a new tab is a bit clunky.

It's not consistently used. I discussed the designs with @kwwii. It was his recommendation to use the icon in the title. I can create a follow up ticket and ask Ken to take a look at it in detail.

kwwii commented:
We should combine everything from Branding, Banners and Home Links in one menu item titled "Customize UI". Users have had problems differentiating between Branding and Banners. Adding another option only makes it worse.

kwwii commented:
We should leave the links in the order the user sets and not automatically order them alphabetically. The default link order should be set according to importance.

kwwii commented:
We can remove the icon from the title. It was a simple attempt to emphasize the fact that the links will open in a new tab/window. 

All inks which are not either a function or lead to another part of the UI itself should open in a new tab/window per default.

kwwii commented:
We can remove the icon from the title. It was a simple attempt to emphasize the fact that the links will open in a new tab/window.

All inks which are not either a function or lead to another part of the UI itself should open in a new tab/window per default.

nwmac reviewed (changes requested):
A few things need fixing:

'Add Label' should be 'Add Link' - and when you click it, the new row should auto-focus the first box
There is no visual feedback when you click 'Restore Defaults' - maybe add a banner at the top to say defaults have been restored?
BUG: The file an issue link does not work if you have not visited the setting page  - the link is empty, which you see when you hover on it - clicking it opens a new dashboard in a new tab
BUG: The links that are within the app cause the page to reload, they should not do that.

Shavindra commented:
> * 'Add Label' should be 'Add Link' - and when you click it, the new row should auto-focus the first box
Fixed label
I am using `KeyValue` component. Issue seems to be with the component. I can follow this up in another ticket

> * There is no visual feedback when you click 'Restore Defaults' - maybe add a banner at the top to say defaults have been restored?
Fixed, timeout set to 10s

https://user-images.githubusercontent.com/1387263/192346764-683120be-8945-4878-b649-c9950bf005bd.mov


> * BUG: The file an issue link does not work if you have not visited the setting page  - the link is empty, which you see when you hover on it - clicking it opens a new dashboard in a new tab

Fixed

> * BUG: The links that are within the app cause the page to reload, they should not do that.
Fixed

> * BUG: We should NOT show the two links '[Find out more about SUSE Rancher Support](https://rancher.com/support-maintenance-terms)' and 'Contact us for pricing' when the user has added a subscription ID. (see: Add 'Commercial Support' link to the 'Links' card - always shown at the bottom unless commercial support enabled on the home page issue)

This will be handled as part of #6665
